front/src/applications/operationalStudies/helpers/TrainSimulationLazyLoader.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  osrdEditoastApi,
+  type PostPacedTrainSimulationSummaryApiResponse,
+  type PostTrainScheduleSimulationSummaryApiResponse,
+  type SimulationSummaryResult,
+} from 'common/api/osrdEditoastApi';
+import type { TimetableItemId } from 'reducers/osrdconf/types';
+import type { AppDispatch } from 'store';
+import {
+  formatEditoastTrainIdToPacedTrainId,
+  formatEditoastTrainIdToTrainScheduleId,
+  formatPacedTrainIdToEditoastTrainId,
+  formatTrainScheduleIdToEditoastTrainId,
+  isTrainSchedule,
+} from 'utils/trainId';
+
+const BATCH_SIZE = 20;
+
+type TrainSimulationLazyLoaderOptions = {
+  dispatch: AppDispatch;
+  infraId: number;
+  electricalProfileSetId?: number;
+  onProgress: (results: Map<TimetableItemId, SimulationSummaryResult>) => void;
+};
+
+/**
+ * Takes a stream of train IDs as input, incrementally invokes onProgress when
+ * train simulations become available.
+ *
+ * This helper only takes care of the API requests, it doesn't perform any
+ * post-processing of the simulation results.
+ */
+export default class TrainSimulationLazyLoader {
+  readonly options: TrainSimulationLazyLoaderOptions;
+
+  pending: TimetableItemId[] = [];
+
+  prevPromise: Promise<void> = Promise.resolve();
+
+  cancelled = false;
+
+  /**
+   * Create a new loader. Options are immutable for the lifetime of the loader.
+   */
+  constructor(options: TrainSimulationLazyLoaderOptions) {
+    this.options = options;
+  }
+
+  /**
+   * Queue train IDs for simulation.
+   */
+  simulateTimetableItems(ids: TimetableItemId[]) {
+    if (this.cancelled) {
+      throw new Error('simulateTimetableItems() called after cancel()');
+    }
+    this.pending = [...this.pending, ...ids];
```

Comment:
here we don’t really care about being immutable either, I would do it in place too.

Reply:
I've changed it to `.push(...ids)` since that doesn't impact readability.